alf/utils/losses.py

```diff
@@ -682,3 +683,132 @@ def forward(self, pred: Tensor, target: Tensor) -> Tensor:
         if ndim > self._batch_dims:
             loss = loss.mean(dim=list(range(self._batch_dims, ndim)))
         return loss
+
+
+class BipartiteMatchingLoss(object):
+    r"""Bipartite matching loss.
+
+    This order-invariant loss can be used to evaluate the matching between a predicted
+    set and a target set. The idea is that for every forward, an optimal one-to-one
+    mapping assignment from the predicted set to the target set is first found using
+    some efficient bipartite graph matching algorithm, and the optimal loss is
+    minimized.
+
+    Mathematically, suppose there are :math:`N` objects in either set,
+    :math:`L(x,y)` is the matching loss between any :math:`(x,y)` object pair,
+    and :math:`\mathcal{G}_N` is the permuation space. The forward loss to be
+    minimized is:
+
+    .. math::
+
+        \min_{g\in\mathcal{G}_N}\sum_n^N L(x_n(\theta),y_{g(n)})
+
+    where :math:`\theta` is the model parameters.
+
+    In practice, to find the optimal assignment, we simply use ``scipy.optimize.linear_sum_assignment``.
+
+    References::
+        `End-to-End Object Detection with Transformers <https://arxiv.org/pdf/2005.12872.pdf>`_, Carion et al.
+
+        `<https://github.com/facebookresearch/detr/blob/main/models/matcher.py>`_
+    """
+
+    def __init__(self,
+                 pair_loss_fn: Callable = torch.cdist,
+                 reduction: str = 'mean',
+                 name: str = "BipartiteMatchingLoss"):
+        """
+        Args:
+            pair_loss_fn: the pairwise matching loss function. It should take
+                two sets of length ``N`` and output a cost matrix of shape ``[N,N]``.
+            reduction: 'sum', 'mean' or 'none'. This is how to reduce the matching
+                loss. For the former two, the loss shape is ``[B]``, while for
+                the 'none', the loss shape is ``[B,N]``.
+        """
+        super().__init__()
+        self._pair_loss_fn = pair_loss_fn
+        self._reduction = reduction
+        assert reduction in ['mean', 'sum', 'none']
+        self._name = name
+
+    def forward(self,
```

Comment:
Why there is a ``forward `` function, ``fast_forward`` and ``__call__``?

Reply:
> same question. What is the difference of forward and fast_forward? Why do we need both of them?

fast_forward constructs a bug cost matrix across the entire batch. The cost matrix is block diagonal so that objects can only be matched to those in the same batch sample. forward uses a for loop to call linear_sum_assignment multiple times, which is about 2x slower.